cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -109,13 +112,100 @@ func (spm *KubeStaticPodPathManager) BackupManifestDir() string {
+
+	// Move the old manifest into the old-manifests directory
+	if err := pathMgr.MoveFile(currentManifestPath, backupManifestPath); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	// Move the new manifest into the manifests directory
+	if err := pathMgr.MoveFile(newManifestPath, currentManifestPath); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	fmt.Printf("[upgrade/staticpods] Moved upgraded manifest to %q and backed up old manifest to %q\n", currentManifestPath, backupManifestPath)
+	fmt.Println("[upgrade/staticpods] Waiting for the kubelet to restart the component")
+
+	// Wait for the mirror Pod hash to change; otherwise we'll run into race conditions here when the kubelet hasn't had time to
+	// notice the removal of the Static Pod, leading to a false positive below where we check that the API endpoint is healthy
+	// If we don't do this, there is a case where we remove the Static Pod manifest, kubelet is slow to react, kubeadm checks the
+	// API endpoint below of the OLD Static Pod component and proceeds quickly enough, which might lead to unexpected results.
+	if err := waiter.WaitForStaticPodControlPlaneHashChange(cfg.NodeName, component, beforePodHash); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	// Wait for the static pod component to come up and register itself as a mirror pod
+	if err := waiter.WaitForPodsWithLabel("component=" + component); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	fmt.Printf("[upgrade/staticpods] Component %q upgraded successfully!\n", component)
+	return nil
+}
 
-	// This string-string map stores the component name and backup filepath (if a rollback is needed).
-	// If a rollback is needed,
+func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathManager, cfg *kubeadmapi.MasterConfiguration, recoverManifests map[string]string) error {
+	// Add etcd static pod spec only if external etcd is not configured
+	if len(cfg.Etcd.Endpoints) != 0 {
+		return fmt.Errorf("external etcd cannot be upgraded with kubeadm")
+	}
+
+	// Need to check currently used version and version from constants, if differs then uprgade
+	runningEtcdVersion, err := util.GetEtcdRunningVersion()
+	if err != nil {
+		return err
+	}
+	targetKubernetesVersion := version.MustParseGeneric(cfg.KubernetesVersion)
+	upgradeEtcdVersion, err := constants.GetEtcdSupportedVersion(uint8(targetKubernetesVersion.Minor()))
+	if err != nil {
+		return err
+	}
+	newEtcdVersion := version.MustParseGeneric(upgradeEtcdVersion)
+	if err != nil {
+		return err
+	}
+	// Ignore error here because the version struct was built with MustParse method
+	// which means any errors will trigger panic.
+	if res, _ := newEtcdVersion.Compare(runningEtcdVersion); res != 1 {
+		return fmt.Errorf("etcd upgrade was requested for the same or more recent version")
```

Comment:
can we either a) return `nil` here to be able to proceed with the control plane upgrade b) return a bool (whether to continue or not) and an error to log only if true is returned?

Reply:
done